### PR TITLE
[FEATURE] Introduce the render cache with cache fragments

### PR DESCRIPTION
**Pull request by e0ipso**
Introduce the render cache at the prepare layer (right before encoding).
Besides, cache fragment entities are declared and created for smart
cache invalidation and fragmentation.

### CODE REVIEW

**Comment by e0ipso** [comment]
This is ready to be reviewed.

I think this is a big leap forward, allowing resource results to be cached for a **LONG** TTL having smart cache invalidation based on cache fragments.

**Comment by amitaibu** [comment]
wow, that's a big one - I'll need to wrap my head around it

**Comment by e0ipso** [comment]
> wow, that's a big one - I'll need to wrap my head around it

want to screenshare for a quick rundown?

**Comment by amitaibu** [comment]
Yeah, I'll go over the code, and then ping. 
btw, tests are failing

**Comment by e0ipso** [comment]
I'm trying to find the time to get back to this. I have not forgotten about it.

**Comment by e0ipso** [comment]
I'll be angry if tests don't come back green.

**Comment by e0ipso** [comment]
:angry:

**Comment by e0ipso** [comment]
Ha! @amitaibu tests are green.

**Comment by amitaibu** [comment]
Just to update that I haven't forgot about this - I'm going to review this tomorrow.
(btw, tests are red)

**Comment by e0ipso** [comment]
> (btw, tests are red)

nope

**Comment by amitaibu** [comment]
sorry for the false alarm, I miss looked ;)

File: restful.entity.inc [hunk 1]
```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
...
+  $query
+    ->entityCondition('entity_type', 'cache_fragment')
+    ->propertyCondition('type', 'entity_type')
+    ->propertyCondition('value', $type);
+  if (!$hashes = CacheFragmentController::lookUpHashes($query)) {
+    return array();
+  }
+
+  // Search for all the cache fragments with our entity id.
+  list($entity_id,,) = entity_extract_ids($type, $entity);
```
Comment (amitaibu) [inline restful.entity.inc]: you can just  `list($entity_id)`

File: restful.entity.inc [hunk 2]
```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
...
  */
 function restful_entity_delete($entity, $type) {
-  list($entity_id) = entity_extract_ids($type, $entity);
-  $cid = 'paet:' . $type . '::ei:' . $entity_id;
-  CacheDecoratedDataProvider::invalidateEntityCache($cid);
+  $resource_manager = restful()->getResourceManager();
+  foreach (_restful_entity_cache_hashes($entity, $type) as $hash) {
+    $handler = $resource_manager->getPlugin(CacheFragmentController::resourceIdFromHash($hash));
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
```
Comment (amitaibu) [inline restful.entity.inc:114]: `// Resource isn't using the default cache decorator.`

File: restful.entity.inc [hunk 3]
```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
...
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
```
Comment (amitaibu) [inline restful.entity.inc:117]: `// Resource takes care of its own invalidation.`

File: restful.entity.inc [hunk 4]
```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
...
+    }
+    $handler
```
Comment (amitaibu) [inline restful.entity.inc:120]: My biggest concern (which isn't huge), is how scalable it is. If we cache per user, and we have 10K users hitting the same resource -- it means we need to delete 10K on a single delete.

Contradicting my own concern - if we have so many caches, one would have to implement it's own invalidation logic.

agree?

File: restful.entity.inc [hunk 5]
```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
...
 
 /**
  * Implements hook_user_update().
  */
 function restful_user_update(&$edit, $account, $category) {
-  // Due to the limitations for cid matching on clearing caches, we need to
-  // clear all the bin. We cannot do $cid = '%::uu' . $account->uid . '::pa';
-  CacheDecoratedDataProvider::invalidateEntityCache('*');
+  // Search for all the cache fragments with our entity id.
+  $query = new \EntityFieldQuery();
```
Comment (amitaibu) [inline restful.entity.inc:131]: why do we need special case of user?

Also wouldn't it run twice - after `restful_entity_update`?

File: restful.entity.inc [hunk 6]
```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
...
+      continue;
+    }
+    if (!$handler instanceof CacheDecoratedResourceInterface) {
+      return;
+    }
+    if (!$handler->hasSimpleInvalidation()) {
+      return;
+    }
+    // You can get away without the fragments for a clear.
+    $cache_object = new RenderCache(new ArrayCollection(), $hash, $handler->getCacheController());
```
Comment (amitaibu) [inline restful.entity.inc:154]: I'm not clear on what we're doing here.

File: restful.entity.inc [hunk 7]
```diff
@@ -29,42 +33,152 @@ function restful_entity_info() {
...
 
 /**
  * Implements hook_user_delete().
  */
 function restful_user_delete($account) {
-  // Due to the limitations for cid matching on clearing caches, we need to
-  // clear all the bin. We cannot do $cid = '%::uu' . $account->uid . '::pa';
-  CacheDecoratedDataProvider::invalidateEntityCache('*');
+  // Search for all the cache fragments with our entity id.
+  $query = new \EntityFieldQuery();
```
Comment (amitaibu) [inline restful.entity.inc:166]: same questions as `restful_user_update()`

File: restful.info [hunk 8]
```diff
@@ -34,7 +34,7 @@ files[] = tests/RestfulListTestCase.test
 ; files[] = tests/RestfulPassThroughTestCase.test
 files[] = tests/RestfulRateLimitTestCase.test
 files[] = tests/RestfulReferenceTestCase.test
-; files[] = tests/RestfulRenderCacheTestCase.test
+files[] = tests/RestfulRenderCacheTestCase.test
```
Comment (amitaibu) [inline restful.info:37]: :sparkling_heart:

File: restful.install [hunk 9]
```diff
@@ -64,6 +64,47 @@ function restful_schema() {
...
+    'description' => 'Cache fragment base table',
+    'fields' => array(
+      'tid' => array(
+        'description' => 'The cache fragment unique ID.',
+        'type' => 'serial',
+        'unsigned' => TRUE,
+        'not null' => TRUE,
+      ),
+      'type' => array(
+        'description' => 'The tag type. Example: entity_id.',
```
Comment (amitaibu) [inline restful.install]: Maybe: `The type of the cache fragment, e.g "entity_id".`

File: restful.module [hunk 10]
```diff
@@ -519,9 +520,11 @@ function restful_restful_resource_alter(ResourceInterface &$resource) {
 
   $plugin_definition = $resource->getPluginDefinition();
   // Check for the rate limit configuration.
-  if (!empty($plugin_definition['renderCache']['render'])) {
-    // NOTE: Disabled cache for now. It needs to be moved to the Render layer.
-    // $resource = new CacheDecoratedResource($resource);
+  if (
```
Comment (amitaibu) [inline restful.module:526]: worth a a comment line.

File: src/Plugin/formatter/Formatter.php [hunk 11]
```diff
@@ -81,4 +84,104 @@ protected static function isIterable($input) {
...
+   * @return bool
+   *   TRUE if the data can be cached.
+   */
+  protected function isCacheEnabled($data) {
+    // We are only caching field collections, but you could cache at different
+    // layers too.
+    if (!$data instanceof ResourceFieldCollectionInterface) {
+      return FALSE;
+    }
+    if (!$context = $data->getContext()) {
```
Comment (amitaibu) [inline src/Plugin/formatter/Formatter.php:102]: why is this needed?

Comment (amitaibu) [inline src/Plugin/formatter/Formatter.php:102]: sorry, I understand it.

File: src/Plugin/formatter/Formatter.php [hunk 12]
```diff
@@ -81,4 +84,104 @@ protected static function isIterable($input) {
...
+   * Gets the cached computed value for the fields to be rendered.
+   *
+   * @param mixed $data
+   *   The data to be rendered.
+   * @param mixed $output
+   *   The rendered data to output.
+   */
+  protected function setCachedData($data, $output) {
+    if (!$render_cache = $this->createCacheController($data)) {
+      return;
```
Comment (amitaibu) [inline src/Plugin/formatter/Formatter.php:134]: just because getCachedData already does -- `return NULL`

File: src/Plugin/formatter/Formatter.php [hunk 13]
```diff
@@ -81,4 +84,104 @@ protected static function isIterable($input) {
...
+   * Gets a cache controller based on the data to be rendered.
+   *
+   * @param mixed $data
+   *   The data to be rendered.
+   *
+   * @return RenderCacheInterface
+   *   The cache controller.
+   */
+  protected function createCacheController($data) {
+    $context = $data->getContext();
```
Comment (amitaibu) [inline src/Plugin/formatter/Formatter.php:149]: use `isCacheEnabled`?

Also, do we really need to do this extra check - wasn't it infact already tested in `isCacheEnabled`?

File: src/Plugin/formatter/Formatter.php [hunk 14]
```diff
@@ -81,4 +84,104 @@ protected static function isIterable($input) {
...
+    if (!$cached_resource = $this->getResource()) {
+      return NULL;
+    }
+    return RenderCache::create($cache_fragments, $cached_resource->getCacheController());
+  }
+
+  /**
+   * Remove the unnecessary fields from the response.
+   *
+   * @param string[] $limit_fields
```
Comment (amitaibu) [inline src/Plugin/formatter/Formatter.php:167]: Is `string[]` a valid phpdocs?

File: src/Plugin/formatter/Formatter.php [hunk 15]
```diff
@@ -81,4 +84,104 @@ protected static function isIterable($input) {
...
+   *   The list of fields.
+   * @param array $output
+   *   The prepared output.
+   *
+   * @return array
+   *   The filtered contents.
+   */
+  public function limitFields($limit_fields, array $output) {
```
Comment (amitaibu) [inline src/Plugin/formatter/Formatter.php]: `array $limit_fields`

File: src/Plugin/formatter/FormatterHalJson.php [hunk 16]
```diff
@@ -186,6 +190,21 @@ protected function extractFieldValues($rows) {
       if (!$rows instanceof ResourceFieldCollectionInterface) {
         throw new InternalServerErrorException('Inconsistent output.');
       }
+
+      // This feels a bit awkward, but if the result is going to be cached, it
+      // pays off the extra effort of generating the whole resource entity. That
+      // way we can get a different field set with the previously cached entity.
+      // If the entity is not going to be cached, then avoid generating the
+      // field data altogether.
+      $limit_fields = $rows->getLimitFields();
```
Comment (amitaibu) [inline src/Plugin/formatter/FormatterHalJson.php:199]: why do even need this check - can't we remove this whole block - and simply wait for the caching [few](https://github.com/RESTful-Drupal/restful/pull/660/files#diff-bf10e1e10ec909f1a0ad72002bb6dd55R221) lines below?

File: src/Plugin/formatter/FormatterJson.php [hunk 17]
```diff
@@ -98,13 +117,17 @@ protected function extractFieldValues($data) {
       }
       $output[$public_field_name] = $value;
     }
+    if ($this->isCacheEnabled($data)) {
```
Comment (amitaibu) [inline src/Plugin/formatter/FormatterJson.php:120]: this duplication is nasty - lets break it down, so formatter implementors don't have to deal with it.

File: src/Plugin/formatter/FormatterJsonApi.php [hunk 18]
```diff
@@ -286,14 +231,80 @@ protected function getRequest() {
...
+      return $this->decompress($output, $included);
+    }
+    foreach ($output['__fields'] as $field_name => $field_contents) {
+      if (empty($field_contents['__embedded'])) {
+        $result['attributes'][$field_name] = $field_contents;
+      }
+      else {
+        // Handle single and multiple relationships.
+        $rel = array();
+        $single_item = $field_contents['__cardinality'] == 1;
```
Comment (amitaibu) [inline src/Plugin/formatter/FormatterJsonApi.php:282]: Damn, I just got a vertigo :wink:

File: src/Plugin/formatter/FormatterJsonApi.php [hunk 19]
```diff
@@ -309,6 +320,142 @@ protected function moveEmbedsToIncluded(array &$output, array $included) {
...
+   *   The prepared output.
+   *
+   * @return array
+   *   The filtered contents.
+   */
+  public function limitFields($limit_fields, array $output) {
+    if (!$limit_fields) {
+      return $output;
+    }
+    if (empty($output['__fields'])) {
```
Comment (amitaibu) [inline src/Plugin/formatter/FormatterJsonApi.php:341]: can't we just pass `$output['__fields']` to the parent method, instead of duplicating this method?

File: src/Plugin/resource/DataProvider/DataProviderEntity.php [hunk 20]
```diff
@@ -231,19 +248,18 @@ public function view($identifier) {
     $id_field_name = empty($this->options['idField']) ? 'id' : $this->options['idField'];
     $resource_field_collection->setIdField($this->fieldDefinitions->get($id_field_name));
 
+    // Defer sparse fieldsets to the formatter. That way we can minimize cache
```
Comment (amitaibu) [inline src/Plugin/resource/DataProvider/DataProviderEntity.php:251]: I don't understand what you mean here

File: src/RenderCache/Entity/CacheFragmentController.php [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+   * Gets the hashes for an EFQ.
+   *
+   * @param \EntityFieldQuery $query
+   *   The EFQ.
+   *
+   * @return string[]
+   *   The hashes that meet the conditions.
+   */
+  public static function lookUpHashes(\EntityFieldQuery $query) {
+    $results = $query->execute();
```
Comment (amitaibu) [inline src/RenderCache/Entity/CacheFragmentController.php:88]: Is there a real use case doing the execute() here, and not getting the results?

**Comment by amitaibu** [comment]
It's official. You are crazy! :)

**Amazing** job

File: src/RenderCache/Entity/CacheFragmentController.php (hunk 21 above)
Comment (e0ipso) [inline src/RenderCache/Entity/CacheFragmentController.php:88]: This is (supposedly) only for developer experience. So they can tweak the query by overriding the controller. Not sure if that is ever gonna happen.

File: restful.entity.inc (hunk 4 above)
Comment (e0ipso) [inline restful.entity.inc:120]: **Agree**

Yeah. This is meant to be just for _simple invalidation_. More complex cases should be handled case by case. That's the reason behind being able to turn off _simple invalidation_.

File: restful.entity.inc (hunk 5 above)
Comment (e0ipso) [inline restful.entity.inc:131]: I asked myself the same question. I'm not sure why it was like this in `1.x`, but I just ported it in this case.

File: restful.entity.inc (hunk 6 above)
Comment (e0ipso) [inline restful.entity.inc:154]: Instead of searching for all the `restful_cache_fragment` entities that reference this cache and delete them + clearing cache, we create the `RenderCache` object (that contains the cache) and call the `->clear()` method on it.

File: src/Plugin/formatter/Formatter.php (hunk 13 above)
Comment (e0ipso) [inline src/Plugin/formatter/Formatter.php:149]: > use isCacheEnabled?

We can't use it, since `isCacheEnabled` only applies to `ResourceFieldCollections`.

> Also, do we really need to do this extra check - wasn't it in fact already tested in isCacheEnabled?

Just making sure, in case we want to use this helper method in other contexts.

File: src/Plugin/formatter/Formatter.php (hunk 14 above)
Comment (e0ipso) [inline src/Plugin/formatter/Formatter.php:167]: yes! Super useful as well for `PhpStorm` type hinting.

File: src/Plugin/formatter/FormatterHalJson.php (hunk 16 above)
Comment (e0ipso) [inline src/Plugin/formatter/FormatterHalJson.php:199]: When caching is enabled and we get a cache MISS we want to generate output for the cache entry for the **whole** entity. That way we can use that cache record independently of the sparse fieldset.

On the other hand, if cache is not enabled we don't want to output for the whole entity, only the bits that we are going to need. For performance reasons.

That is why we bail just before `->render()`. `->render()` does the heavy lifting to get eh field info.

File: src/Plugin/formatter/FormatterJson.php (hunk 17 above)
Comment (e0ipso) [inline src/Plugin/formatter/FormatterJson.php:120]: I tried (unsuccessfully) to find an elegant way to do this. The problem is that we don't have only one way to generate the output, since it will depend on runtime configuration.

Comment (e0ipso) [inline src/Plugin/formatter/FormatterJson.php:120]: Ideas?

File: src/Plugin/formatter/FormatterJsonApi.php (hunk 18 above)
Comment (e0ipso) [inline src/Plugin/formatter/FormatterJsonApi.php:282]: Yeah, there is a lot of _proprietary format_ business logic here. I did not know how to avoid this and still have cacheable entities (with all the includes that they may need in the cache record).

This denormalization + optional caching + re-normalization is (IMO) the ugliest part of RESTful. I can see that we can improve this in future sweeps, but I think it's a _good enough_ **first** iteration.

File: src/Plugin/formatter/FormatterJsonApi.php [hunk 22]
```diff
@@ -286,14 +231,80 @@ protected function getRequest() {
...
    *
+   * @return array
+   *   The processed data.
+   *
    * @throws \Drupal\restful\Exception\InternalServerErrorException
    */
-  protected function moveEmbedsToIncluded(array &$output, array $included) {
+  protected function decompress(array $output, array &$included) {
+    static $depth = -1;
+    $depth++;
```
Comment (e0ipso) [inline src/Plugin/formatter/FormatterJsonApi.php:251]: Not sure how I feel myself about this `$depth` variable. It's certainly useful for debugging purposes (I use it all the time for conditional breakpoints), but it doesn't drive any business logic.

File: src/Plugin/formatter/FormatterJsonApi.php (hunk 19 above)
Comment (e0ipso) [inline src/Plugin/formatter/FormatterJsonApi.php:341]: We'd need to do the `empty` check in the parent method in order to avoid PHP warnings.

File: src/Plugin/resource/DataProvider/DataProviderEntity.php (hunk 20 above)
Comment (e0ipso) [inline src/Plugin/resource/DataProvider/DataProviderEntity.php:251]: I tried to explain https://github.com/RESTful-Drupal/restful/pull/660/files#r41708622

Does it make sense that way?

**Comment by e0ipso** [comment]
Thank you so much for the (in)sanity check.

I really think that we needed to improve performance, while making sure that we don't get cache poisoning and we have all the tools for smart invalidation. To me having _yet another_ TTL-based cache layer didn't make a lot of sense, since we already have Drupal's page cache, Varnish (and Akamai) doing that. I wanted the render cache entries to be able to be cached **forever** and never be cleared, except by the invalidation methods.

**Comment by e0ipso** [comment]
## Next steps

### Making sure that we don't have infinite recursion https://github.com/RESTful-Drupal/restful/issues/676

When entity `A` references `B` and in turn `B` references `C, D and A`, we need to replace the embedded `A` by a placeholder that gets expanded in `->decompress`.

### Making sure we add cache fragments for embedded entities https://github.com/RESTful-Drupal/restful/issues/677

When entity `A` references `B` and in turn `B` references `C, D and A`. If the cache entry for `D` gets updated we need to clear the caches of `D`, `B` and `A`.

### Improve performance on uncached requests https://github.com/RESTful-Drupal/restful/issues/678

Right now we are fetching the embedded information for relationships without checking if the relationship was requested using the `include` query parameter.

File: restful.entity.inc (hunk 1 above)
Comment (e0ipso) [inline restful.entity.inc]: :+1:

File: restful.entity.inc (hunk 2 above)
Comment (e0ipso) [inline restful.entity.inc:114]: :+1: 

`// Resource isn't using a cache decorator.`

File: restful.entity.inc (hunk 3 above)
Comment (e0ipso) [inline restful.entity.inc:117]: :+1:

File: restful.install (hunk 9 above)
Comment (e0ipso) [inline restful.install]: :+1:

File: restful.module (hunk 10 above)
Comment (e0ipso) [inline restful.module:526]: :+1:

File: src/Plugin/formatter/Formatter.php (hunk 12 above)
Comment (e0ipso) [inline src/Plugin/formatter/Formatter.php:134]: In general if you don't have a `@return` tag in your docblock, the recommendation is to return _void_ instead of _NULL_. Even if those are the same in PHP.

Returning different from `return;` in a function that does not return a value will make PhpStorm unhappy.

File: src/Plugin/formatter/Formatter.php (hunk 15 above)
Comment (e0ipso) [inline src/Plugin/formatter/Formatter.php]: :+1:

**Comment by e0ipso** [comment]
@amitaibu feedback has been applied.

File: restful.entity.inc (hunk 5 above)
Comment (amitaibu) [inline restful.entity.inc:131]: I think It's because of the granularity of the cache clear - https://github.com/RESTful-Drupal/restful/blob/f0f981a8552a8c8b429bce2991d06e551661adff/restful.entity.inc#L56-L58

File: restful.entity.inc (hunk 6 above)
Comment (amitaibu) [inline restful.entity.inc:154]: `$cache_object->clear();` is just a proxy to actually delete all the fragments, no? If so `// You can get away without the fragments for a clear.` isn't very clear

File: src/Plugin/formatter/Formatter.php (hunk 12 above)
Comment (amitaibu) [inline src/Plugin/formatter/Formatter.php:134]: ok

File: src/Plugin/formatter/FormatterHalJson.php (hunk 16 above)
Comment (amitaibu) [inline src/Plugin/formatter/FormatterHalJson.php:199]: :+1:

File: src/Plugin/formatter/FormatterJson.php (hunk 17 above)
Comment (amitaibu) [inline src/Plugin/formatter/FormatterJson.php:120]: not very beautiful, but maybe have the main method broken to three parts? -- implementors will have to deal only with the middle part

File: src/Plugin/formatter/FormatterJsonApi.php (hunk 18 above)
Comment (amitaibu) [inline src/Plugin/formatter/FormatterJsonApi.php:282]: it's fine -- I just admitted my review of this part was poorly done ;)

File: src/Plugin/formatter/FormatterJsonApi.php (hunk 19 above)
Comment (amitaibu) [inline src/Plugin/formatter/FormatterJsonApi.php:341]: matter of taste - for me duplication of method is more extreme then `!empty()` - but I can go either way

File: src/Plugin/resource/DataProvider/DataProviderEntity.php (hunk 20 above)
Comment (amitaibu) [inline src/Plugin/resource/DataProvider/DataProviderEntity.php:251]: Sure, maybe just add a `@see` with a reference example?

File: src/RenderCache/Entity/CacheFragmentController.php (hunk 21 above)
Comment (amitaibu) [inline src/RenderCache/Entity/CacheFragmentController.php:88]: > Not sure if that is ever gonna happen.

Me neither, but it's ok to keep as is I guess.

**Comment by amitaibu** [comment]
Minor comments + failing tests. After that it can be merged :)

File: restful.entity.inc (hunk 5 above)
Comment (e0ipso) [inline restful.entity.inc:131]: Then we don't need this anymore!

**Comment by e0ipso** [comment]
Merging as per https://github.com/RESTful-Drupal/restful/pull/660#issuecomment-147382355

**Comment by amitaibu** [comment]
Cool stuff!
